goss-testing/scripts/data-validator.py

```diff
@@ -0,0 +1,289 @@
+  else:
+    try:
+      print_err(f"\nRunning on node: {hostname}. Querying BSS...")
+      print_err("------------------------------------------------------------")
+      command = [ "cray", "bss", "bootparameters", "list", "--format", "json" ]
+      bss_proc = subprocess.Popen(command, stdout=subprocess.PIPE)
+      return json.loads(bss_proc.stdout.read())
+    except Exception as e:
+      print_err(str(e))
+      return 1
+
+
+def user_data(data):
+  """Returns the 'user-data' blobs from data.json or BSS.
+  """
+  filtered_data = []
+
+  # if we're using BSS data
+  if isinstance(data, list):
+    for blob in data:
+        if isinstance(blob['cloud-init']['user-data'], dict):
+          filtered_data.append(blob['cloud-init']['user-data'])
+
+  # if we're using basecamp data
+  elif isinstance(data, dict) and "Global" in data:
+    for blob in data:
+      if blob != "Global":
+        filtered_data.append(data[blob]['user-data'])
+  
+  return filtered_data
+
+
+def drill_down_data(data, desired_keys):
+  for i in range(0, ( len(desired_keys) + 1), 1):
+    if i == len(desired_keys):
+      return saved_data
+    if i == 0:
+      saved_data = data[desired_keys[i]]
+    elif i != 0:
+      saved_data = saved_data[desired_keys[i]]
+
+
+def are_valid_ip_masks(data, desired_keys):
+  """Checks that IP/mask or CIDR is valid.
+  """
+  filtered_data = user_data(data)
+  err = 0
+
+  for blob in filtered_data:
+    instance_hostname = blob['hostname']
+    child_key = drill_down_data(blob, desired_keys)
+
+    if not child_key:
+      print_err(f"ERR: {desired_keys} is not defined for: {instance_hostname}")
+      err = 1
+    else:
+      for value in child_key:
+        try:
+          ipaddress.IPv4Network(value, strict=False)
+        except ValueError:
+          print_err(f"ERR: {instance_hostname}: '{value}' is not a valid IP/mask in {desired_keys}")
+          err = 1
+  if err == 1:
+    return err
+
+def check_hostname_syntax(hostname):
+  """Checks that a given hostname syntax is valid.
+  """  
+  if len(hostname) > 253:
+      return False
+  if hostname[-1] == ".":
+      hostname = hostname[:-1]
+  allowed = re.compile("(?!-)[A-Z\d-]{1,63}(?<!-)$", re.IGNORECASE)
+  return all(allowed.match(x) for x in hostname.split("."))
+
+
+def are_valid_hostnames(data, desired_keys):
+  """Checks that a list of hostnames is defined and valid.
+  """
+  filtered_data = user_data(data)
+  err = 0
+
+  for blob in filtered_data:
+    instance_hostname = blob['hostname']
+    child_key = drill_down_data(blob, desired_keys)
+
+    if not child_key:
+      print_err(f"ERR: {desired_keys} is not defined for: {instance_hostname}: ")
+      err = 1
+    else:
+      for item in drill_down_data(blob, desired_keys):
+        if not check_hostname_syntax(item):
+          print_err(f"ERR: {instance_hostname}: '{item}' is not a valid hostname in {desired_keys}")
+          err = 1
+  return err
+
+
+def are_hosts_sane(data, desired_keys):
```

Comment:
This function passes if the data (or the filtered_data) is empty. Is this intended? Note -- I don't know the answer here. If it's okay for the data to be missing, and we're just wanting to make sure it's valid if it is present, then the current behavior is okay.

Reply:
I'm including a check for empty data and a check for confirming that desired_keys exist.